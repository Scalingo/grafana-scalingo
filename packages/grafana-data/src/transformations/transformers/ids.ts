export enum DataTransformerID {
  //  join = 'join', // Pick a field and merge all series based on that field
  append = 'append', // Merge all series together
  //  rotate = 'rotate', // Columns to rows
  reduce = 'reduce', // Run calculations on fields

  filterFields = 'filterFields', // Pick some fields (keep all frames)
  filterFieldsByName = 'filterFieldsByName', // Pick fields with name matching regex (keep all frames)
  filterFrames = 'filterFrames', // Pick some frames (keep all fields)
  filterByRefId = 'filterByRefId', // Pick some frames by RefId
  noop = 'noop', // Does nothing to the dataframe
}
