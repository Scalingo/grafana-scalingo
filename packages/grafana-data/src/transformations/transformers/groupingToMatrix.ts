import { map } from 'rxjs/operators';

import { MutableDataFrame } from '../../dataframe';
import { getFieldDisplayName } from '../../field/fieldState';
import { DataFrame, DataTransformerInfo, Field, FieldType, Vector } from '../../types';

import { DataTransformerID } from './ids';

export interface GroupingToMatrixTransformerOptions {
  columnField?: string;
  rowField?: string;
  valueField?: string;
}

const DEFAULT_COLUMN_FIELD = 'Time';
const DEFAULT_ROW_FIELD = 'Time';
const DEFAULT_VALUE_FIELD = 'Value';

export const groupingToMatrixTransformer: DataTransformerInfo<GroupingToMatrixTransformerOptions> = {
  id: DataTransformerID.groupingToMatrix,
  name: 'Grouping to Matrix',
  description: 'Groups series by field and return a matrix visualisation',
  defaultOptions: {
    columnField: DEFAULT_COLUMN_FIELD,
    rowField: DEFAULT_ROW_FIELD,
    valueField: DEFAULT_VALUE_FIELD,
  },

  operator: (options) => (source) =>
    source.pipe(
      map((data) => {
        const columnFieldMatch = options.columnField || DEFAULT_COLUMN_FIELD;
        const rowFieldMatch = options.rowField || DEFAULT_ROW_FIELD;
        const valueFieldMatch = options.valueField || DEFAULT_VALUE_FIELD;

        // Accept only single queries
        if (data.length !== 1) {
          return data;
        }

        const frame = data[0];
        const keyColumnField = findKeyField(frame, columnFieldMatch);
        const keyRowField = findKeyField(frame, rowFieldMatch);
        const valueField = findKeyField(frame, valueFieldMatch);
        const rowColumnField = `${rowFieldMatch}\\${columnFieldMatch}`;

        if (!keyColumnField || !keyRowField || !valueField) {
          return data;
        }

        const columnValues = uniqueValues(keyColumnField.values);
        const rowValues = uniqueValues(keyRowField.values);

        const matrixValues: { [key: string]: { [key: string]: any } } = {};

        for (let index = 0; index < valueField.values.length; index++) {
          const columnName = keyColumnField.values.get(index);
          const rowName = keyRowField.values.get(index);
          const value = valueField.values.get(index);

          if (!matrixValues[columnName]) {
            matrixValues[columnName] = {};
          }

          matrixValues[columnName][rowName] = value;
        }

        const resultFrame = new MutableDataFrame();

        resultFrame.addField({
          name: rowColumnField,
          values: rowValues,
          type: FieldType.string,
        });

        for (const columnName of columnValues) {
          let values = [];
          for (const rowName of rowValues) {
            const value = matrixValues[columnName][rowName] ?? '';
            values.push(value);
          }

          resultFrame.addField({
            name: columnName.toString(),
            values: values,
            config: valueField.config,
            type: valueField.type,
          });
        }

        return [resultFrame];
      })
    ),
};

function uniqueValues(values: Vector): any[] {
  const unique = new Set();

  for (let index = 0; index < values.length; index++) {
    unique.add(values.get(index));
  }

  return Array.from(unique);
}

function findKeyField(frame: DataFrame, matchTitle: string): Field | null {
  for (let fieldIndex = 0; fieldIndex < frame.fields.length; fieldIndex++) {
    const field = frame.fields[fieldIndex];

    if (matchTitle === getFieldDisplayName(field)) {
      return field;
    }
  }

  return null;
}
