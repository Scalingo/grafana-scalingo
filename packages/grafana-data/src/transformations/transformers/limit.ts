import { map } from 'rxjs/operators';

import { DataTransformerInfo } from '../../types';
import { ArrayVector } from '../../vector/ArrayVector';

import { DataTransformerID } from './ids';

export interface LimitTransformerOptions {
  limitField?: number;
}

const DEFAULT_LIMIT_FIELD = 10;

export const limitTransformer: DataTransformerInfo<LimitTransformerOptions> = {
  id: DataTransformerID.limit,
  name: 'Limit',
  description: 'Limit the number of items to the top N',
  defaultOptions: {
    limitField: DEFAULT_LIMIT_FIELD,
  },

  operator: (options) => (source) =>
    source.pipe(
      map((data) => {
        const limitFieldMatch = options.limitField || DEFAULT_LIMIT_FIELD;
        return data.map((frame) => {
          if (frame.length > limitFieldMatch) {
            return {
              ...frame,
              fields: frame.fields.map((f) => {
                const vals = f.values.toArray();
                return {
                  ...f,
                  values: new ArrayVector(vals.slice(0, limitFieldMatch)),
                };
              }),
              length: limitFieldMatch,
            };
          }

          return frame;
        });
      })
    ),
};
