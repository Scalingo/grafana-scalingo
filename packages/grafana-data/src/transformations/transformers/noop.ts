import { DataTransformerID } from './ids';
import { DataFrame } from '../../types/dataFrame';
import { DataTransformerInfo } from '../../types/transformations';

export interface NoopTransformerOptions {
  include?: string;
  exclude?: string;
}

export const noopTransformer: DataTransformerInfo<NoopTransformerOptions> = {
  id: DataTransformerID.noop,
  name: 'noop',
  description: 'No-operation transformer',
  defaultOptions: {},

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: NoopTransformerOptions) => {
    return (data: DataFrame[]) => data;
  },
};
