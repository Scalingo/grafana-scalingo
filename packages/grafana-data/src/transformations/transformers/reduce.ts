import { DataTransformerID } from './ids';
import { MatcherConfig, DataTransformerInfo } from '../../types/transformations';
import { ReducerID, fieldReducers, reduceField } from '../fieldReducer';
import { alwaysFieldMatcher } from '../matchers/predicates';
import { DataFrame, Field, FieldType } from '../../types/dataFrame';
import { ArrayVector } from '../../vector/ArrayVector';
import { KeyValue } from '../../types/data';
import { guessFieldTypeForField } from '../../dataframe/processDataFrame';
import { getFieldMatcher } from '../matchers';

export interface ReduceTransformerOptions {
  reducers: ReducerID[];
  fields?: MatcherConfig; // Assume all fields
}

export const reduceTransformer: DataTransformerInfo<ReduceTransformerOptions> = {
  id: DataTransformerID.reduce,
  name: 'Reducer',
  description: 'Return a DataFrame with the reduction results',
  defaultOptions: {
    reducers: [ReducerID.min, ReducerID.max, ReducerID.mean, ReducerID.last],
  },

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: ReduceTransformerOptions) => {
    const matcher = options.fields ? getFieldMatcher(options.fields) : alwaysFieldMatcher;
    const calculators = options.reducers && options.reducers.length ? fieldReducers.list(options.reducers) : [];
    const reducers = calculators.map(c => c.id);

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        const values: ArrayVector[] = [];
        const fields: Field[] = [];
        const byId: KeyValue<ArrayVector> = {};
        values.push(new ArrayVector()); // The name
        fields.push({
          name: 'Field',
          type: FieldType.string,
          values: values[0],
          config: {},
        });
        for (const info of calculators) {
          const vals = new ArrayVector();
          byId[info.id] = vals;
          values.push(vals);
          fields.push({
            name: info.id,
            type: FieldType.other, // UNKNOWN until after we call the functions
            values: values[values.length - 1],
            config: {
              title: info.name,
              // UNIT from original field?
            },
          });
        }
        for (let i = 0; i < series.fields.length; i++) {
          const field = series.fields[i];
          if (matcher(field)) {
            const results = reduceField({
              field,
              reducers,
            });
            // Update the name list
            values[0].buffer.push(field.name);
            for (const info of calculators) {
              const v = results[info.id];
              byId[info.id].buffer.push(v);
            }
          }
        }
        for (const f of fields) {
          const t = guessFieldTypeForField(f);
          if (t) {
            f.type = t;
          }
        }
        processed.push({
          ...series, // Same properties, different fields
          fields,
          length: values[0].length,
        });
      }
      return processed;
    };
  },
};
