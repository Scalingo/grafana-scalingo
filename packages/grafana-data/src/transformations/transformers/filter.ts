import { noopTransformer } from './noop';
import { DataFrame, Field } from '../../types/dataFrame';
import { DataTransformerID } from './ids';
import { DataTransformerInfo, MatcherConfig } from '../../types/transformations';
import { FieldMatcherID } from '../matchers/ids';
import { getFieldMatcher, getFrameMatchers } from '../matchers';

export interface FilterOptions {
  include?: MatcherConfig;
  exclude?: MatcherConfig;
}

export const filterFieldsTransformer: DataTransformerInfo<FilterOptions> = {
  id: DataTransformerID.filterFields,
  name: 'Filter Fields',
  description: 'select a subset of fields',
  defaultOptions: {
    include: { id: FieldMatcherID.numeric },
  },

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: FilterOptions) => {
    if (!options.include && !options.exclude) {
      return noopTransformer.transformer({});
    }

    const include = options.include ? getFieldMatcher(options.include) : null;
    const exclude = options.exclude ? getFieldMatcher(options.exclude) : null;

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        // Find the matching field indexes
        const fields: Field[] = [];
        for (let i = 0; i < series.fields.length; i++) {
          const field = series.fields[i];
          if (exclude) {
            if (exclude(field)) {
              continue;
            }
            if (!include) {
              fields.push(field);
            }
          }
          if (include && include(field)) {
            fields.push(field);
          }
        }

        if (!fields.length) {
          continue;
        }
        const copy = {
          ...series, // all the other properties
          fields, // but a different set of fields
        };
        processed.push(copy);
      }
      return processed;
    };
  },
};

export const filterFramesTransformer: DataTransformerInfo<FilterOptions> = {
  id: DataTransformerID.filterFrames,
  name: 'Filter Frames',
  description: 'select a subset of frames',
  defaultOptions: {},

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: FilterOptions) => {
    if (!options.include && !options.exclude) {
      return noopTransformer.transformer({});
    }

    const include = options.include ? getFrameMatchers(options.include) : null;
    const exclude = options.exclude ? getFrameMatchers(options.exclude) : null;

    return (data: DataFrame[]) => {
      const processed: DataFrame[] = [];
      for (const series of data) {
        if (exclude) {
          if (exclude(series)) {
            continue;
          }
          if (!include) {
            processed.push(series);
          }
        }
        if (include && include(series)) {
          processed.push(series);
        }
      }
      return processed;
    };
  },
};
