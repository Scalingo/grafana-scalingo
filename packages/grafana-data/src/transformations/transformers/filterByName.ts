import { DataTransformerID } from './ids';
import { filterFieldsTransformer, FilterOptions } from './filter';
import { DataTransformerInfo } from '../../types/transformations';
import { FieldMatcherID } from '../matchers/ids';

export interface FilterFieldsByNameTransformerOptions {
  include?: string;
  exclude?: string;
}

export const filterFieldsByNameTransformer: DataTransformerInfo<FilterFieldsByNameTransformerOptions> = {
  id: DataTransformerID.filterFieldsByName,
  name: 'Filter fields by name',
  description: 'select a subset of fields',
  defaultOptions: {},

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: FilterFieldsByNameTransformerOptions) => {
    const filterOptions: FilterOptions = {};
    if (options.include) {
      filterOptions.include = {
        id: FieldMatcherID.byName,
        options: options.include,
      };
    }
    if (options.exclude) {
      filterOptions.exclude = {
        id: FieldMatcherID.byName,
        options: options.exclude,
      };
    }

    return filterFieldsTransformer.transformer(filterOptions);
  },
};
