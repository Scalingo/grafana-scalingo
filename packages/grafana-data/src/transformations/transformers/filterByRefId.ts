import { DataTransformerID } from './ids';
import { FilterOptions, filterFramesTransformer } from './filter';
import { DataTransformerInfo } from '../../types/transformations';
import { FrameMatcherID } from '../matchers/ids';

export interface FilterFramesByRefIdTransformerOptions {
  include?: string;
  exclude?: string;
}

export const filterFramesByRefIdTransformer: DataTransformerInfo<FilterFramesByRefIdTransformerOptions> = {
  id: DataTransformerID.filterByRefId,
  name: 'Filter data by query refId',
  description: 'select a subset of results',
  defaultOptions: {},

  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transformer: (options: FilterFramesByRefIdTransformerOptions) => {
    const filterOptions: FilterOptions = {};
    if (options.include) {
      filterOptions.include = {
        id: FrameMatcherID.byRefId,
        options: options.include,
      };
    }
    if (options.exclude) {
      filterOptions.exclude = {
        id: FrameMatcherID.byRefId,
        options: options.exclude,
      };
    }

    return filterFramesTransformer.transformer(filterOptions);
  },
};
