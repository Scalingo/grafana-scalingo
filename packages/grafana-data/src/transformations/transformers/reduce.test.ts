import { ReducerID } from '../fieldReducer';
import { DataTransformerID } from './ids';
import { toDataFrame, toDataFrameDTO } from '../../dataframe/processDataFrame';
import { transformDataFrame } from '../transformers';

const seriesWithValues = toDataFrame({
  fields: [
    { name: 'A', values: [1, 2, 3, 4] }, // Numbers
    { name: 'B', values: ['a', 'b', 'c', 'd'] }, // Strings
  ],
});

describe('Reducer Transformer', () => {
  it('filters by include', () => {
    const cfg = {
      id: DataTransformerID.reduce,
      options: {
        reducers: [ReducerID.first, ReducerID.min, ReducerID.max, ReducerID.delta],
      },
    };
    const processed = transformDataFrame([cfg], [seriesWithValues])[0];
    expect(processed.fields.length).toBe(5);
    expect(toDataFrameDTO(processed)).toMatchSnapshot();
  });
});
