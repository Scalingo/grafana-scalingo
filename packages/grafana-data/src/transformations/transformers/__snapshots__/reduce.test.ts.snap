// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer Transformer filters by include 1`] = `
Object {
  "fields": Array [
    Object {
      "config": Object {},
      "labels": undefined,
      "name": "Field",
      "type": "string",
      "values": Array [
        "A",
        "B",
      ],
    },
    Object {
      "config": Object {
        "title": "First",
      },
      "labels": undefined,
      "name": "first",
      "type": "number",
      "values": Array [
        1,
        "a",
      ],
    },
    Object {
      "config": Object {
        "title": "Min",
      },
      "labels": undefined,
      "name": "min",
      "type": "number",
      "values": Array [
        1,
        null,
      ],
    },
    Object {
      "config": Object {
        "title": "Max",
      },
      "labels": undefined,
      "name": "max",
      "type": "number",
      "values": Array [
        4,
        null,
      ],
    },
    Object {
      "config": Object {
        "title": "Delta",
      },
      "labels": undefined,
      "name": "delta",
      "type": "number",
      "values": Array [
        3,
        0,
      ],
    },
  ],
  "meta": Object {
    "transformations": Array [
      "reduce",
    ],
  },
  "name": undefined,
  "refId": undefined,
}
`;
